File: 3sum.cpp
// TLE
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // dictionary of result. the key is a vector of three numbers, the value is a boolean
        map<vector<int>, bool> results;

        
        for (int i = 0; i < nums.size() - 2; i++) {
            for (int j = i + 1; j < nums.size() - 1; j++) {
                for (int k = j + 1; k < nums.size(); k++) {
                    if (nums[i] + nums[j] + nums[k] == 0) {
                        // add to results if it doesn't already exist. sort the num[i], num[j], num[k] first
                        vector<int> result = {nums[i], nums[j], nums[k]};
                        sort(result.begin(), result.end());
                        results[result] = true;
                    }
                }
            }
        }
        // convert the map to a vector
        vector<vector<int>> mapResults;

        for (auto it = results.begin(); it != results.end(); it++) {
            mapResults.push_back(it->first);
        }
        return mapResults;
    }
};

File: 3sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # create dictionary of num and it's index
        result = []
        nums.sort()
        dictionaryOfNum = {}
        for index, num in enumerate(nums):
            dictionaryOfNum[num] = index
        testedI = {}

        dictonaryOfResult = {}
        
        for i in range(0, len(nums)):
            if nums[i] in testedI.keys():
                continue
            testedI[nums[i]] = 1
            for j in range(i + 1, len(nums)):
                # find in dict
                if(-(nums[i] + nums[j])) in dictionaryOfNum.keys():
                    # check if index is larger than j
                    k = dictionaryOfNum[-(nums[i] + nums[j])]
                    if(k > j):
                        newRes = [nums[i], nums[j], nums[k]]
                        newResStr= ''.join(str(x) for x in newRes)
                        if newResStr not in dictonaryOfResult.keys():
                            result.append(newRes)
                            dictonaryOfResult[newResStr] = 1               
        
        return result
                       
        


sol = Solution()
res = sol.threeSum([1,-1,-1,0])
print(res)

